internal/testrand/extreme.go

```diff
@@ -0,0 +1,205 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package testrand
+
+import (
+	"math"
+
+	"golang.org/x/exp/rand"
+)
+
+// extreme is a pseudo-random number generator that has high probability of returning extreme values.
+type extreme struct {
+	probability    float64
+	nanProbability float64
+
+	rnd Rand
+}
+
+// newExtreme creates a new extreme pseudo-random generator.
+// p is the probability of returning an extreme value.
+// nan is the probability of returning a NaN.
+func newExtreme(p, nan float64, rnd Rand) *extreme {
+	return &extreme{p, nan, rnd}
+}
+
+// Perm returns a permutation of integers [0, n).
+func (e *extreme) Perm(n int) []int { return e.rnd.Perm(n) }
+
+// Read generates len(p) pseudo-random bytes.
+func (e *extreme) Read(p []byte) (n int, err error) { return e.rnd.Read(p) }
+
+// Seed reseeds the pseudo-random generator.
+func (e *extreme) Seed(seed uint64) { e.rnd.Seed(seed) }
+
+// Shuffle shuffles n items using the swap callback.
+func (e *extreme) Shuffle(n int, swap func(i, j int)) { e.rnd.Shuffle(n, swap) }
+
+// p returns true when the generator should output an extreme value.
+func (e *extreme) p() bool {
+	if e.probability <= 0 {
+		return false
+	}
+	return e.rnd.Float64() < e.probability
+}
+
+// nan returns true when the generator should output nan.
+func (e *extreme) nan() bool {
+	if e.nanProbability <= 0 {
+		return false
+	}
+	return e.rnd.Float64() < e.nanProbability
+}
+
+// ExpFloat64 returns an exponentialy distributed pseudo-random float64 in range (0, math.MaxFloat64].
+func (e *extreme) ExpFloat64() float64 {
+	if e.p() {
```

Comment:
according to https://golang.org/doc/effective_go.html#switch, it's idiomatic to write `if/else-if` with a `switch {}`.
```go
switch {
case e.p():
    // ...
case e.nan():
    // ...
}
```